### PR TITLE
Breaking: Properly Handle Instruction Scheduling

### PR DESCRIPTION
**Pull request by kalzoo**
- Implement `NONBLOCKING` parsing
- Correctly implement frame blocking/utilitization on `Program#get_frames_for_instruction`, replacing the misnamed `Program#get_blocking_frames`.
- Fix the parsing of `RAW-CAPTURE`, `CAPTURE`
- Add parsing for `FENCE`
- Fix display for several instructions including `DELAY`
- Correctly express instruction dependencies on frames
- Support DOT format export for `InstructionBlock` and `ScheduledProgram` to aid in program debugging
- Add snapshot testing and several snapshot tests to assert that dependencies are working as expected
- Fix how labels are auto-generated when `ScheduledProgram` cuts a program into blocks; prevent creation of empty blocks with no label
- Fix expression parsing to allow memory references (although the spec has ambiguity here)

TODO:
- [x] Tests for frame blocking
- [x] `CAPTURE` and memory dependencies

### CODE REVIEW

**Comment by kalzoo** [comment]
Diffing and debugging program scheduling dependencies is way easier now, thanks to [insta](https://docs.rs/insta/1.7.1/insta/).

### What it looked like before, when something about program block parsing and labeling changed

(these two examples are just illustrative, it's not the same test case)

```
---- program::graph::tests::blocks::with_control_flow stdout ----
{"block-0": InstructionBlock { instructions: [Pulse { blocking: true, frame: FrameIdentifier { name: "rx", qubits: [Fixed(0)] }, waveform: WaveformInvocation { name: "test", parameters: {"duration": Number(Complex { re: 0.0000009999999974752427, im: 0.0 })} } }, Pulse { blocking: true, frame: FrameIdentifier { name: "rx", qubits: [Fixed(0)] }, waveform: WaveformInvocation { name: "test", parameters: {"duration": Number(Complex { re: 0.0000009999999974752427, im: 0.0 })} } }, Delay { duration: Number(Complex { re: 1.0, im: 0.0 }), frame_names: [], qubits: [Fixed(0)] }], graph: {BlockStart: [(InstructionIndex(0), Outgoing), (BlockEnd, Outgoing)], InstructionIndex(0): [(BlockStart, Incoming), (InstructionIndex(1), Outgoing)], InstructionIndex(1): [(InstructionIndex(0), Incoming), (InstructionIndex(2), Outgoing)], InstructionIndex(2): [(InstructionIndex(1), Incoming), (BlockEnd, Outgoing)], BlockEnd: [(BlockStart, Incoming), (InstructionIndex(2), Incoming)]}, terminator: Continue }, "part2": InstructionBlock { instructions: [Pulse { blocking: true, frame: FrameIdentifier { name: "rx", qubits: [Fixed(0)] }, waveform: WaveformInvocation { name: "test", parameters: {"duration": Number(Complex { re: 0.0000009999999974752427, im: 0.0 })} } }, Delay { duration: Number(Complex { re: 2.0, im: 0.0 }), frame_names: [], qubits: [Fixed(0)] }], graph: {BlockStart: [(InstructionIndex(0), Outgoing), (BlockEnd, Outgoing)], InstructionIndex(0): [(BlockStart, Incoming), (InstructionIndex(1), Outgoing)], InstructionIndex(1): [(InstructionIndex(0), Incoming), (BlockEnd, Outgoing)], BlockEnd: [(BlockStart, Incoming), (InstructionIndex(1), Incoming)]}, terminator: Continue }, "part3": InstructionBlock { instructions: [Delay { duration: Number(Complex { re: 3.0, im: 0.0 }), frame_names: [], qubits: [Fixed(0)] }], graph: {BlockStart: [(InstructionIndex(0), Outgoing), (BlockEnd, Outgoing)], InstructionIndex(0): [(BlockStart, Incoming), (BlockEnd, Outgoing)], BlockEnd: [(BlockStart, Incoming), (InstructionIndex(0), Incoming)]}, terminator: Unconditional { target: "part2" } }, "block-1": InstructionBlock { instructions: [Delay { duration: Number(Complex { re: 4.0, im: 0.0 }), frame_names: [], qubits: [Fixed(0)] }], graph: {BlockStart: [(InstructionIndex(0), Outgoing), (BlockEnd, Outgoing)], InstructionIndex(0): [(BlockStart, Incoming), (BlockEnd, Outgoing)], BlockEnd: [(BlockStart, Incoming), (InstructionIndex(0), Incoming)]}, terminator: Unconditional { target: "block5" } }, "block-2": InstructionBlock { instructions: [], graph: {BlockStart: [(BlockEnd, Outgoing)], BlockEnd: [(BlockStart, Incoming)]}, terminator: Halt }, "block5": InstructionBlock { instructions: [Delay { duration: Number(Complex { re: 5.0, im: 0.0 }), frame_names: [], qubits: [Fixed(0)] }], graph: {BlockStart: [(InstructionIndex(0), Outgoing), (BlockEnd, Outgoing)], InstructionIndex(0): [(BlockStart, Incoming), (BlockEnd, Outgoing)], BlockEnd: [(BlockStart, Incoming), (InstructionIndex(0), Incoming)]}, terminator: Halt }}
thread 'program::graph::tests::blocks::with_control_flow' panicked at 'assertion failed: `(left == right)`
  left: `6`,
 right: `7`', quil/src/program/graph.rs:739:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

🤨


### What it looks like now

(even better in a terminal with color support)

```
---- program::graph::tests::graph::single_instruction stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Differences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: quil/src/program/snapshots/quil__program__graph__tests__graph__single_instruction.snap
Snapshot: single_instruction
Source: quil/src/program/graph.rs:795
-old snapshot
+new results
──────────────────────────────────────────────────────────────────────────────────────
ProgramDebugWrapper {
    program: &scheduled_program,
}
────────────┬─────────────────────────────────────────────────────────────────────────
    0     0 │ digraph {
    1     1 │   entry [label="Entry Point"]
    2       │-  entry -> "start_start"
    3       │-  subgraph "cluster_start" {
    4       │-          label="start"
          2 │+  entry -> "block-0_start"
          3 │+  subgraph "cluster_block-0" {
          4 │+          label="block-0"
    5     5 │           node [ style="filled" ]
    6       │-          "start_start" [ label=start, shape=circle ]
    7       │-          "start_start" -> "start_0" [ label="immediate" ]
    8       │-          "start_start" -> "start_end" [ label="immediate" ]
    9       │-          "start_0" [label="PULSE 0 \"rx\" test(duration: 1.00e6)"]
   10       │-          "start_0" -> "start_end" [ label="immediate" ]
   11       │-          "start_end" [ label=end, shape=circle ]
          6 │+          "block-0_start" [ label=start, shape=circle ]
          7 │+          "block-0_start" -> "block-0_0" [ label="immediate" ]
          8 │+          "block-0_start" -> "block-0_end" [ label="immediate" ]
          9 │+          "block-0_0" [label="PULSE 0 \"rx\" test(duration: 1.00e6)"]
         10 │+          "block-0_0" -> "block-0_end" [ label="immediate" ]
         11 │+          "block-0_end" [ label=end, shape=circle ]
   12    12 │   }
   13    13 │ }
────────────┴─────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
```

Better yet, you can take those snapshot files and drop them in to a graphviz renderer like [this one online](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%09entry%20%5Blabel%3D%22Entry%20Point%22%5D%0A%09entry%20-%3E%20%22block_0_start%22%0A%09subgraph%20%22cluster_block_0%22%20%7B%0A%09%09label%3D%22block_0%22%0A%09%09node%20%5B%20style%3D%22filled%22%20%5D%0A%09%09%22block_0_start%22%20%5B%20label%3Dstart%2C%20shape%3Dcircle%20%5D%0A%09%09%22block_0_start%22%20-%3E%20%22block_0_0%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22block_0_start%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22block_0_0%22%20%5Blabel%3D%22PULSE%200%20%5C%22rf%5C%22%20test(duration%3A%201.00e6)%22%5D%0A%09%09%22block_0_0%22%20-%3E%20%22block_0_1%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22block_0_1%22%20%5Blabel%3D%22PULSE%200%20%5C%22rf%5C%22%20test(duration%3A%201.00e6)%22%5D%0A%09%09%22block_0_1%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22block_0_end%22%20%5B%20label%3Dend%2C%20shape%3Dcircle%20%5D%0A%09%7D%0A%7D) and get something that's easy to understand. This will help not just when we're debugging changes in the tests, but also for folks debugging complex Quil programs.

Here's another example: [single-qubit active reset](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%09entry%20%5Blabel%3D%22Entry%20Point%22%5D%0A%09entry%20-%3E%20%22measure_start%22%0A%09subgraph%20%22cluster_measure%22%20%7B%0A%09%09label%3D%22measure%22%0A%09%09node%20%5B%20style%3D%22filled%22%20%5D%0A%09%09%22measure_start%22%20%5B%20label%3Dstart%2C%20shape%3Dcircle%20%5D%0A%09%09%22measure_start%22%20-%3E%20%22measure_0%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22measure_start%22%20-%3E%20%22measure_1%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22measure_start%22%20-%3E%20%22measure_end%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22measure_0%22%20%5Blabel%3D%22NONBLOCKING%20PULSE%200%20%5C%22ro_tx%5C%22%20test(duration%3A%201.00e6)%22%5D%0A%09%09%22measure_0%22%20-%3E%20%22measure_end%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22measure_1%22%20%5Blabel%3D%22NONBLOCKING%20CAPTURE%200%20%5C%22ro_rx%5C%22%20test(duration%3A%201.00e6)%20ro%5B0%5D%22%5D%0A%09%09%22measure_1%22%20-%3E%20%22measure_end%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22measure_end%22%20%5B%20label%3Dend%2C%20shape%3Dcircle%20%5D%0A%09%7D%0A%22measure_end%22%20-%3E%20%22end_start%22%20%5Blabel%3D%22if%20ro%5B0%5D%20%3D%3D%200%22%5D%0A%22measure_end%22%20-%3E%20%22feedback_start%22%20%5Blabel%3D%22if%20ro%5B0%5D%20!%3D%200%22%5D%0A%09subgraph%20%22cluster_feedback%22%20%7B%0A%09%09label%3D%22feedback%22%0A%09%09node%20%5B%20style%3D%22filled%22%20%5D%0A%09%09%22feedback_start%22%20%5B%20label%3Dstart%2C%20shape%3Dcircle%20%5D%0A%09%09%22feedback_start%22%20-%3E%20%22feedback_0%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22feedback_start%22%20-%3E%20%22feedback_end%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22feedback_0%22%20%5Blabel%3D%22PULSE%200%20%5C%22rf%5C%22%20test(duration%3A%201.00e6)%22%5D%0A%09%09%22feedback_0%22%20-%3E%20%22feedback_end%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22feedback_end%22%20%5B%20label%3Dend%2C%20shape%3Dcircle%20%5D%0A%09%7D%0A%22feedback_end%22%20-%3E%20%22measure_start%22%20%5Blabel%3D%22always%22%5D%0A%09subgraph%20%22cluster_end%22%20%7B%0A%09%09label%3D%22end%22%0A%09%09node%20%5B%20style%3D%22filled%22%20%5D%0A%09%09%22end_start%22%20%5B%20label%3Dstart%2C%20shape%3Dcircle%20%5D%0A%09%09%22end_start%22%20-%3E%20%22end_end%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22end_end%22%20%5B%20label%3Dend%2C%20shape%3Dcircle%20%5D%0A%09%7D%0A%7D).

image

**Comment by kalzoo** [comment]
This now includes memory dependencies: see [a sample graph](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%09entry%20%5Blabel%3D%22Entry%20Point%22%5D%0A%09entry%20-%3E%20%22block_0_start%22%0A%09subgraph%20%22cluster_block_0%22%20%7B%0A%09%09label%3D%22block_0%22%0A%09%09node%20%5B%20style%3D%22filled%22%20%5D%0A%09%09%22block_0_start%22%20%5B%20label%3Dstart%2C%20shape%3Dcircle%20%5D%0A%09%09%22block_0_start%22%20-%3E%20%22block_0_0%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22block_0_0%22%20%5Blabel%3D%22MOVE%20a%5B0%5D%201%22%5D%0A%09%09%22block_0_0%22%20-%3E%20%22block_0_1%22%20%5B%20label%3D%22immediate%22%20%5D%0A%09%09%22block_0_0%22%20-%3E%20%22block_0_2%22%20%5B%20label%3D%22await%20write%22%20%5D%0A%09%09%22block_0_1%22%20%5Blabel%3D%22MOVE%20b%5B0%5D%202%22%5D%0A%09%09%22block_0_1%22%20-%3E%20%22block_0_2%22%20%5B%20label%3D%22await%20write%22%20%5D%0A%09%09%22block_0_2%22%20%5Blabel%3D%22ADD%20a%5B0%5D%20b%5B0%5D%22%5D%0A%09%09%22block_0_2%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22await%20write%22%20%5D%0A%09%09%22block_0_end%22%20%5B%20label%3Dend%2C%20shape%3Dcircle%20%5D%0A%09%7D%0A%7D%0A%0A).

I'm not entirely settled on the implementation - I'm not sure it exposes the right level of abstraction to the library's consumer. What does the user need to know about why one instruction "depends" on another?

Previously, there were two variants:

* "Immediate" - can execute as soon as precursor is complete but no sooner
* "Await Capture" - can execute after allowing the capture value to propagate into memory. This felt non-general and quite backend-specific.

So now "Immediate" is unchanged but "await capture" has become "await memory access" which is quite likely unnecessarily complex. I'm not sure why a user needs to know that instruction A's read must complete before B's write can begin.

On the other hand, "immediate" is still abstract, instead of providing the _reason_ why it's "immediate". The two reasons are: stable ordering of classical instructions in the order they appear in the program; and using a common frame.

So "immediate" is abstract but "await memory access" is detailed, and that doesn't seem right either.

**Comment by kalzoo** [comment]
So after some thought I settled on "more detail" as the cleaner approach in terms of execution dependencies, and I like the results. Graph edges now express zero or more dependencies in a `HashSet`, rather than just one as before. This means that one instruction can depend on another for multiple reasons, not just one (because `petgraph::Graph` is not a multigraph, dependencies were overwriting one another.) This also makes one of the clumiest things in this PR go away - the requirement to maintain precedences among different dependency types to make sure that an "await capture" is not overwritten by an "await read," which could cause incorrect program compilation later on.

Take a look at [this graph](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%09entry%20%5Blabel%3D%22Entry%20Point%22%5D%0A%09entry%20-%3E%20%22block_0_start%22%0A%09subgraph%20%22cluster_block_0%22%20%7B%0A%09%09label%3D%22block_0%22%0A%09%09node%20%5B%20style%3D%22filled%22%20%5D%0A%09%09%22block_0_start%22%20%5B%20label%3Dstart%2C%20shape%3Dcircle%20%5D%0A%09%09%22block_0_start%22%20-%3E%20%22block_0_0%22%20%5B%20label%3D%22frame%22%20%5D%0A%09%09%22block_0_start%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22ordering%22%20%5D%0A%09%09%22block_0_0%22%20%5Blabel%3D%22CAPTURE%200%20%5C%22ro_rx%5C%22%20test(a%3A%20param%5B0%5D)%20ro%5B0%5D%22%5D%0A%09%09%22block_0_0%22%20-%3E%20%22block_0_1%22%20%5B%20label%3D%22await%20capture%0Aframe%22%20%5D%0A%09%09%22block_0_0%22%20-%3E%20%22block_0_2%22%20%5B%20label%3D%22await%20capture%0Aframe%22%20%5D%0A%09%09%22block_0_0%22%20-%3E%20%22block_0_3%22%20%5B%20label%3D%22await%20capture%0Aframe%22%20%5D%0A%09%09%22block_0_0%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22await%20read%22%20%5D%0A%09%09%22block_0_1%22%20%5Blabel%3D%22NONBLOCKING%20PULSE%200%20%5C%22rf%5C%22%20test(a%3A%20ro%5B0%5D)%22%5D%0A%09%09%22block_0_1%22%20-%3E%20%22block_0_3%22%20%5B%20label%3D%22await%20read%0Aframe%22%20%5D%0A%09%09%22block_0_1%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22await%20read%22%20%5D%0A%09%09%22block_0_2%22%20%5Blabel%3D%22NONBLOCKING%20PULSE%201%20%5C%22rf%5C%22%20test(a%3A%20ro%5B0%5D)%22%5D%0A%09%09%22block_0_2%22%20-%3E%20%22block_0_3%22%20%5B%20label%3D%22await%20read%0Aframe%22%20%5D%0A%09%09%22block_0_2%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22await%20read%22%20%5D%0A%09%09%22block_0_3%22%20%5Blabel%3D%22CAPTURE%200%20%5C%22ro_rx%5C%22%20test(a%3A%20param%5B0%5D)%20ro%5B0%5D%22%5D%0A%09%09%22block_0_3%22%20-%3E%20%22block_0_4%22%20%5B%20label%3D%22await%20capture%0Aframe%22%20%5D%0A%09%09%22block_0_3%22%20-%3E%20%22block_0_5%22%20%5B%20label%3D%22await%20capture%0Aframe%22%20%5D%0A%09%09%22block_0_3%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22await%20capture%0Aawait%20read%0Aframe%22%20%5D%0A%09%09%22block_0_4%22%20%5Blabel%3D%22NONBLOCKING%20PULSE%200%20%5C%22rf%5C%22%20test(a%3A%20ro%5B0%5D)%22%5D%0A%09%09%22block_0_4%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22await%20read%0Aframe%22%20%5D%0A%09%09%22block_0_5%22%20%5Blabel%3D%22NONBLOCKING%20PULSE%201%20%5C%22rf%5C%22%20test(a%3A%20ro%5B0%5D)%22%5D%0A%09%09%22block_0_5%22%20-%3E%20%22block_0_end%22%20%5B%20label%3D%22await%20read%0Aframe%22%20%5D%0A%09%09%22block_0_end%22%20%5B%20label%3Dend%2C%20shape%3Dcircle%20%5D%0A%09%7D%0A%7D%0A%0A): many of the edges have multiple dependencies, and I think it expresses the model quite well. If someone is wondering why the timing of their program is not as expected, they can examine this graph and see exactly what logic is causing the sequencing of instructions; "immediate" told no such story.

File: Cargo.toml [hunk 1]
```diff
@@ -12,3 +12,6 @@ nom = "6.1.0"
 num-complex = "0.3.0"
 petgraph = "0.5.1"
 serde = { version = "1.0.125", features = ["derive"] }
+
+[dev-dependencies]
```
Comment (genos) [inline Cargo.toml:16]: Might users of the library want access to `insta`-generated things, too?

File: src/parser/command.rs [hunk 2]
```diff
@@ -134,7 +139,9 @@ pub fn parse_defwaveform<'a>(input: ParserInput<'a>) -> ParserResult<'a, Instruc
         token!(RParenthesis),
     ))(input)?;
     let parameters = parameters.unwrap_or_default();
-    let (input, sample_rate) = token!(Float(v))(input)?;
+    // NOTE: this cast will cause unexpected behavior for very large integers that cannot be accurately cast to f64.
```
Comment (genos) [inline src/parser/command.rs:142]: Would it be worth checking for that and outputting an `Err`?

File: src/parser/common.rs [hunk 3]
```diff
@@ -201,3 +210,34 @@ pub fn skip_newlines_and_comments<'a>(input: ParserInput<'a>) -> ParserResult<'a
...
+                        name: "ro".to_owned(),
+                        index: 0
+                    })
+                )
+            ]
+            .into_iter()
+            .collect()
+        )
+    }
+}
```
Comment (genos) [inline src/parser/common.rs:243]: Nicely tested, thanks!

I was contemplating trying to add some property-based testing library like [`proptest`](https://github.com/AltSysrq/proptest) to generate round-trip tests like `assert_eq!(program, parse(to_string(program))`. An idea for a different PR.

File: src/parser/instruction.rs [hunk 4]
```diff
@@ -399,7 +436,20 @@ mod tests {
...
                 })
-            }
+            },
+            Instruction::Pulse {
+                blocking: true,
+                frame: FrameIdentifier {
+                    name: "xy".to_owned(),
+                    qubits: vec![Qubit::Fixed(0)]
+                },
+                waveform: Box::new(WaveformInvocation {
```
Comment (genos) [inline src/parser/instruction.rs:446]: Does the `WaveformInvocation` take up enough space that it should be heap-allocated, ergo `Box::new`?

File: src/parser/macros.rs [hunk 5]
```diff
@@ -62,6 +62,22 @@ macro_rules! token {
             }
         }
     }};
+    ($expected_variant: ident($enm:ident::$variant:ident)) => {{
```
Comment (genos) [inline src/parser/macros.rs:65]: `expected_variant`

Loki_(TV_series)_logo
variant

File: src/program/snapshots/quil__program__graph__tests__graph__active_reset_single_frame.snap [hunk 6]
```diff
@@ -0,0 +1,44 @@
+---
```
Comment (genos) [inline src/program/snapshots/quil__program__graph__tests__graph__active_reset_single_frame.snap:1]: 👏 👏  These are great!

**Review by genos — COMMENTED** [review]
Nice work! A few questions here & there, but no burning concerns

File: Cargo.toml (hunk 1 above)
Comment (kalzoo) [reply Cargo.toml:16]: Nope - these are just snapshots of test outputs

File: src/parser/common.rs (hunk 3 above)
Comment (kalzoo) [reply src/parser/common.rs:243]: Sure, I'd be interested in seeing what that'd look like here

File: src/parser/instruction.rs (hunk 4 above)
Comment (kalzoo) [reply src/parser/instruction.rs:446]: Ah, there must have been a self-referencing type here at some point in the past, but I don't see a need for it now. I'll make this its own issue.

Comment (kalzoo) [reply src/parser/instruction.rs:446]: Issue: https://github.com/rigetti/quil-rust/issues/15

Comment (genos) [reply src/parser/instruction.rs:446]: Cheers, I'll put together a happy little PR.

File: src/parser/command.rs (hunk 2 above)
Comment (kalzoo) [reply src/parser/command.rs]: Took a second to grok what `nom` wanted here, but this has now been fixed

Comment (genos) [reply src/parser/command.rs]: Nice!

**Comment by genos** [comment]
👍 nice work @kalzoo

**Review by genos — APPROVED** [review]
👍 well done!

**Pull request merged**